### PR TITLE
Do not pass default properties to access control checks

### PR DESCRIPTION
**Pull request by kokosing**
Do not pass default properties to access control checks

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java [hunk 1]
```diff
@@ -119,7 +127,14 @@ public ListenableFuture<Void> execute(
                 Optional.empty(),
                 properties);
 
-        accessControl.checkCanCreateMaterializedView(session.toSecurityContext(), name, properties);
+        Set<String> specifiedPropertyKeys = statement.getProperties().stream()
+                // property names are case-insensitive and normalized to lower case
+                .map(property -> property.getName().getValue().toLowerCase(ENGLISH))
+                .collect(toImmutableSet());
+        Map<String, Object> explicitlySetProperties = properties.keySet().stream()
+                .filter(specifiedPropertyKeys::contains)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java:135]: lowercase here when invoking `contains`

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 2]
```diff
@@ -81,9 +98,12 @@
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
+@Test(singleThreaded = true)
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:101]: Add a comment why

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 3]
```diff
@@ -94,14 +114,27 @@ protected QueryRunner createQueryRunner()
...
+                    @Override
+                    protected SystemAccessControl delegate()
+                    {
+                        return systemAccessControl.get();
+                    }
+                })
                 .build();
         queryRunner.installPlugin(new BlackHolePlugin());
-        queryRunner.createCatalog("blackhole", "blackhole");
+        queryRunner.createCatalog("blackhole", "blackhole", Map.of());
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:126]: why the change?

File: core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java (hunk 1 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java:135]: It is already lower cased in `io.trino.metadata.PropertyUtil#evaluateProperties`

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java (hunk 3 above)
Comment (kokosing) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:126]: I was thinking to remove this plugin from this test. However I failed and I failed to revert changes, as you can see,

File: core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java:135]: Can this class depend on that? maybe it should not, so that the code is "more obviously correct".
i.e. consisten with lowercase 3 lines up

Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java:132]: Also `io.trino.spi.session.PropertyMetadata#PropertyMetadata` disallows to create a property with non lower cased property name.

**Comment by martint** [comment]
What’s the motivation for this change? If the values of default properties are not passed, how would the access control know what value they are supposed to take?

**Comment by kokosing** [comment]
I am sorry, I was under impression that the change is straightforward so I gave very limited commit message.

The motivation is to make access control for table properties to be more practical and coherent with access for other objects.

>  If the values of default properties are not passed, how would the access control know what value they are supposed to take?

In my opinion access control does not need to know or control what engine is doing in order to fulfill user requests, but it is more about controlling what users are doing.

As example consider the below (that may happen with existing implementation):
```
CREATE TABLE t(i int);
ERROR: Access denied: cannot set location to 's3://bucket/data/schema/t'
```
In such situation, I believe users don't care about location of where data is stored. They may not even know that used catalog is a Hive connector. They just want to create a table. They have CREATE TABLE privilege and they ask engine to do whatever it needs, but please create a table for me. User didn't ask for any specific location. 

Few more issues with existing implementation:
 - authorization services (like Ranger or file based ones) typically deny everything by default. In such case, if they would like provide access control for table properties, then admins need to control each table property. Even the hidden ones and ones that are even mentioned in any query by users of the system.
 - "Ok, but I can grant access to all table properties to bypass the above". Yes but if you would like to deny certain usage of table properties then you need to still enumerate all other table properties to grant access to them so user can CREATE TABLE.
 - Considering the above, whenever we add new properties admins will be forced to update they authorization policies to upgrade Trino.
 - Hive has 21 table properties now,  it if access control requires network communication for each of resource access then it may require some time to do so.
 - if access control is tracking audit access log, then simple CREATE TABLE for Hive connector will produce 21 additional events. Which could be surprising because as nobody really used these table properties.
 - it leaks hidden table properties to access control (and possibly to audit services)

The logic behind proposed implementation:
 - if admin creates a catalog with certain defaults then they say it is ok to CREATE TABLES using these defaults. So anybody who has CREATE TABLE privilege can create them. However, admins can simply disallow to overriding defaults or control which table properties are allowed to be modified by users.
 - it does not expose mentioned issues with existing implementation for access control of table properties
 - proposed implementation is coherent with how we control session properties. We call access control only when somebody requests `SET SESSION...`, not when we populate a session.

**Comment by kokosing** [comment]
Updated tests a bit, to make sure not used properties do not leak.

@martint @findepi Any more comments?

File: core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/CreateMaterializedViewTask.java:132]: > Also `io.trino.spi.session.PropertyMetadata#PropertyMetadata` disallows to create a property with non lower cased property name.

today.

if you rely on that other class guarantees, add a `verify` here (`.peek(...)`)

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 4]
```diff
@@ -756,4 +818,112 @@ public void testUseStatementAccessControlWithDeniedSchema()
...
+        @Override
+        public void checkCanSetMaterializedViewProperties(SystemSecurityContext context, CatalogSchemaTableName materializedView, Map<String, Optional<Object>> properties)
+        {
+            checkProperties(properties);
+        }
+
+        private static void checkProperties(Map<?, ?> properties)
+        {
+            if (!properties.isEmpty()) {
+                throw new AccessDeniedException("Cannot access properties: " + properties.keySet());
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:925]: do you have any test that would try to set 2+ properties?

i think it would be nice to have
but then, use of keySet could cause test flakiness (indeterminism), unless we eg sort

**Review by findepi — APPROVED** [review]

**Comment by colebow** [comment]
Does this need a release note? @kokosing

**Comment by kokosing** [comment]
I don't think so.
